src/group_theory/free_abelian_group.lean

```diff
@@ -277,6 +307,44 @@ instance : is_comm_applicative free_abelian_group.{u} :=
       (λ y₁ y₂ ih1 ih2, by rw [map_add, map_add, add_seq, ih1, ih2]))
     (λ p ih, by rw [map_neg, neg_seq, seq_neg, ih])
     (λ x₁ x₂ ih1 ih2, by rw [map_add, add_seq, seq_add, ih1, ih2]) }
+
+
+end monad
+
+universe w
+
+variables {β : Type v} {γ : Type w}
+
+/-- The additive group homomorphism `free_abelian_group α →+ free_abelian_group β` induced from a
+  map `α → β` -/
+def map (f : α → β) : free_abelian_group α →+ free_abelian_group β :=
+lift (of ∘ f)
+
+lemma lift_comp {α} {β} {γ} [add_comm_group γ]
+  (f : α → β) (g : β → γ) (x : free_abelian_group α) :
+  lift (g ∘ f) x = lift g (map f x) :=
+begin
+  apply free_abelian_group.induction_on x,
+  { exact add_monoid_hom.map_zero _ },
+  { intro y, refl },
+  { intros x h, simp only [h, add_monoid_hom.map_neg] },
+  { intros x y h₁ h₂, simp only [h₁, h₂, add_monoid_hom.map_add] }
+end
+
+lemma map_id : map id = add_monoid_hom.id (free_abelian_group α) :=
+eq.symm $ lift.ext _ _ $ λ x, lift.unique of (add_monoid_hom.id _) $ λ y, add_monoid_hom.id_apply _
+
+lemma map_id_apply (x : free_abelian_group α) : map id x = x := by {rw map_id, refl }
+
+lemma map_comp {f : α → β} {g : β → γ} : map (g ∘ f) = (map g).comp (map f) :=
+eq.symm $ lift.ext _ _ $ λ x, eq.symm $ lift_comp _ _ _
+
+lemma map_comp_apply {f : α → β} {g : β → γ} (x : free_abelian_group α) :
+  map (g ∘ f) x = (map g) ((map f) x) := by { rw map_comp, refl }
+
+-- version of map_of which uses `map`
+lemma map_of' {f : α → β} (a : α) : map f (of a) = of (f a) := rfl
+
```

Comment:
There are some theorems in the `monad` section that have `map` in their name but talk about `<$>` and the theorems here that talk about `map`. Will this lead to confusion later?

Is it worth proving something about how this `map` is related to `<$>`?

(These are questions for anyone with more familiarity with monad-ology in mathlib than me.)

Reply:
The <$> thing was just Kenny learning about monads IIRC. I am happy to have them in this file but they're not polymorphic, and also in this case the map is a group homomorphism; my impression is that generally we reserve the "standard keywords" like `lift` and `map` for the homomorphisms rather than the bare maps. I agree that there could be confusion. One way of removing the confusion would be by deleting all the monad nonsense completely, but it seemed to me that most of them were never going to be used by people wanting to do algebra -- `map_of` won't rewrite and `map_add` is accessible via dot notation from the add_monoid_hom map. I could call this one `map_of_apply` and make a note in the module docstring explaining what's going on.

Reply:
Thanks for the explanation! OK, if the "monad nonsense" is unlikely to be used, then I think the homomorphisms should get the nicer / more convenient names, if possible. Either way, a note would be helpful.

Reply:
@kbuzzard, what do you think we should do here? If you think getting rid of the monad stuff is the right approach, I could have a go at that.

Otherwise, perhaps just add a comment explaining the current situation, and we'll merge?

Reply:
Sorry, I need to focus on my course right now. My plan was to write something about what's going on in the implementation issues part of the module docstring.

Reply:
OK I have explained what is going on with this `map` shenannigans in the module docstring.